### PR TITLE
testing with mocking homework

### PR DESCRIPTION
**Pull request by randy525**

### CODE REVIEW

**Comment by icovdii** [comment]
Make PR name verbose, please.

File: src/test/java/com/endava/internship/mocking/repository/InMemPaymentRepositoryTest.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import java.util.Optional;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+public class InMemPaymentRepositoryTest {
+
+    InMemUserRepository userRepository;
```
Comment (icovdii) [inline src/test/java/com/endava/internship/mocking/repository/InMemPaymentRepositoryTest.java]: All fields must be private final, always.

File: src/test/java/com/endava/internship/mocking/repository/InMemPaymentRepositoryTest.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+    @Test
+    void findById() {
+        User expectedUser = new User(1, "John", Status.ACTIVE);
+
+        Exception exception = assertThrows(IllegalArgumentException.class, () -> userRepository.findById(null));
+        assertEquals(exception.getMessage(), "User id must not be null");
+
+        assertEquals(userRepository.findById(0), Optional.empty());
+        assertEquals(userRepository.findById(1), Optional.of(expectedUser));
```
Comment (icovdii) [inline src/test/java/com/endava/internship/mocking/repository/InMemPaymentRepositoryTest.java]: Good test, just one comment: you try to check here 2 different test cases: negative and positive. Usually for every case we create separate test. Do it, please, and give to tests verbose names, don't worry, name can be very long.

File: src/test/java/com/endava/internship/mocking/service/BasicValidationServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+import com.endava.internship.mocking.model.Status;
+import com.endava.internship.mocking.model.User;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+public class BasicValidationServiceTest {
+
+    BasicValidationService validationService;
```
Comment (icovdii) [inline src/test/java/com/endava/internship/mocking/service/BasicValidationServiceTest.java]: final

File: src/test/java/com/endava/internship/mocking/service/BasicValidationServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    @BeforeEach
+    void init() {
+        validationService = new BasicValidationService();
+    }
+    @Test
+    public void validateAmount() {
+        Exception exception = assertThrows(IllegalArgumentException.class, () -> validationService.validateAmount(null));
+        assertEquals(exception.getMessage(), "Amount must not be null");
+        exception = assertThrows(IllegalArgumentException.class, () -> validationService.validateAmount(-5.7));
+        assertEquals(exception.getMessage(), "Amount must be greater than 0");
```
Comment (icovdii) [inline src/test/java/com/endava/internship/mocking/service/BasicValidationServiceTest.java:35]: Same, you check here 2 different test cases, separate them

File: src/test/java/com/endava/internship/mocking/service/BasicValidationServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        Exception exception = assertThrows(IllegalArgumentException.class, () -> validationService.validateUser(user));
+        assertEquals(exception.getMessage(), "User with id " + user.getId() + " not in ACTIVE status");
+    }
+
+    @Test
+    public void validateMessage() {
+        Exception exception = assertThrows(IllegalArgumentException.class, () -> validationService.validateMessage(null));
+        assertEquals(exception.getMessage(), "Payment message must not be null");
+    }
+}
```
Comment (icovdii) [inline src/test/java/com/endava/internship/mocking/service/BasicValidationServiceTest.java]: for 4 methods above add tests for happy cases, validation with correct data

File: src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java [hunk 6]
```diff
@@ -1,31 +1,75 @@
...
+import java.util.Optional;
+import java.util.UUID;
 
 @ExtendWith(MockitoExtension.class)
 class PaymentServiceTest {
 
+    ValidationService validationServiceMock ;
+    UserRepository userRepositoryMock;
+    PaymentRepository paymentRepositoryMock;
+    PaymentService paymentService;
```
Comment (icovdii) [inline src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java]: finalize all fields

File: src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java [hunk 7]
```diff
@@ -1,31 +1,75 @@
...
+        verify(validationServiceMock).validateAmount(150.5);
+        verify(validationServiceMock).validateUserId(0);
+        verify(validationServiceMock).validateUser(expectedUser);
+
+        verify(userRepositoryMock).findById(0);
+        verify(paymentRepositoryMock).save(argumentCaptor.capture());
+
+        assertEquals(argumentCaptor.getValue().getUserId(), expectedUser.getId());
+        assertEquals(argumentCaptor.getValue().getAmount(), 150.5);
+        assertTrue(argumentCaptor.getValue().getMessage().contains(expectedUser.getName()));
```
Comment (icovdii) [inline src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java:54]: Good test. To make it perfect you can add verifyNoMoreInteractions() at the end, generally, add it at the end of every test for every mock. And try to understand why it is needed :) Or may be you know it yet.

File: src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java [hunk 8]
```diff
@@ -1,31 +1,75 @@
...
     }
```
Comment (icovdii) [inline src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java:59]: And add if possible unhappy testcase.

File: src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java [hunk 9]
```diff
@@ -1,31 +1,75 @@
...
     void setUp() {
+        validationServiceMock = mock(ValidationService.class);
+        userRepositoryMock = mock(UserRepository.class);
+        paymentRepositoryMock = mock(PaymentRepository.class);
+        paymentService = new PaymentService(userRepositoryMock, paymentRepositoryMock, validationServiceMock);
 
     }
 
     @Test
     void createPayment() {
```
Comment (icovdii) [inline src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java]: more verbose name, please

File: src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java [hunk 10]
```diff
@@ -1,31 +1,75 @@
...
+        verify(userRepositoryMock).findById(0);
+        verify(paymentRepositoryMock).save(argumentCaptor.capture());
+
+        assertEquals(argumentCaptor.getValue().getUserId(), expectedUser.getId());
+        assertEquals(argumentCaptor.getValue().getAmount(), 150.5);
+        assertTrue(argumentCaptor.getValue().getMessage().contains(expectedUser.getName()));
     }
 
     @Test
     void editMessage() {
```
Comment (icovdii) [inline src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java]: give more verbose names to tests

File: src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java [hunk 11]
```diff
@@ -1,31 +1,75 @@
...
+        UUID paymentID = UUID.randomUUID();
+        String newString = "Any String";
+        paymentService.editPaymentMessage(paymentID, newString);
+
+        verify(validationServiceMock).validatePaymentId(paymentID);
+        verify(validationServiceMock).validateMessage(newString);
 
+        verify(paymentRepositoryMock).editMessage(paymentID, newString);
```
Comment (icovdii) [inline src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java:71]: Add here some asserts if possible.

File: src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java [hunk 12]
```diff
@@ -1,31 +1,75 @@
...
     }
 
     @Test
     void getAllByAmountExceeding() {
-
+        final double amount = 13.37;
+        paymentService.getAllByAmountExceeding(amount);
+        verify(paymentRepositoryMock).findAll();
```
Comment (icovdii) [inline src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java:99]: In 90% of test should be asserts, so add some asserts here please.

**Review by icovdii — CHANGES_REQUESTED** [review]

File: src/test/java/com/endava/internship/mocking/repository/InMemPaymentRepositoryTest.java (hunk 2 above)
Comment (randy525) [reply src/test/java/com/endava/internship/mocking/repository/InMemPaymentRepositoryTest.java]: done

File: src/test/java/com/endava/internship/mocking/service/BasicValidationServiceTest.java (hunk 3 above)
Comment (randy525) [reply src/test/java/com/endava/internship/mocking/service/BasicValidationServiceTest.java]: same answer, I can't make it final

File: src/test/java/com/endava/internship/mocking/service/BasicValidationServiceTest.java (hunk 4 above)
Comment (randy525) [reply src/test/java/com/endava/internship/mocking/service/BasicValidationServiceTest.java:35]: done

File: src/test/java/com/endava/internship/mocking/service/BasicValidationServiceTest.java (hunk 5 above)
Comment (randy525) [reply src/test/java/com/endava/internship/mocking/service/BasicValidationServiceTest.java]: done

File: src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java (hunk 7 above)
Comment (randy525) [reply src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java:54]: done

File: src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java (hunk 8 above)
Comment (randy525) [reply src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java:59]: I found only one test that can be added

File: src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java (hunk 9 above)
Comment (randy525) [reply src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java]: done

File: src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java (hunk 10 above)
Comment (randy525) [reply src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java]: done

File: src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java (hunk 12 above)
Comment (randy525) [reply src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java:99]: done

File: src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java (hunk 11 above)
Comment (randy525) [reply src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java:71]: iI haven't thought of anything to add here:)

File: src/test/java/com/endava/internship/mocking/service/BasicValidationServiceTest.java [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+        validationService = new BasicValidationService();
+    }
+
+    @Test
+    public void validateAmount_HappyCase() {
+        assertDoesNotThrow(() -> validationService.validateAmount(100.0));
+    }
+
+    @Test
+    public void validateAmountWhenPutNull() {
```
Comment (icovdii) [inline src/test/java/com/endava/internship/mocking/service/BasicValidationServiceTest.java:27]: Good name.

File: src/test/java/com/endava/internship/mocking/service/BasicValidationServiceTest.java [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+        Exception exception = assertThrows(IllegalArgumentException.class, () -> validationService.validateAmount(null));
+        assertEquals(exception.getMessage(), "Amount must not be null");
+    }
+
+    @Test
+    public void validateAmountWhenPutNegativeNumber() {
```
Comment (icovdii) [inline src/test/java/com/endava/internship/mocking/service/BasicValidationServiceTest.java:33]: Good name, well done

File: src/test/java/com/endava/internship/mocking/repository/InMemPaymentRepositoryTest.java (hunk 1 above)
Comment (icovdii) [reply src/test/java/com/endava/internship/mocking/repository/InMemPaymentRepositoryTest.java]: Ok, leave it as it is.

File: src/test/java/com/endava/internship/mocking/service/BasicValidationServiceTest.java (hunk 3 above)
Comment (icovdii) [reply src/test/java/com/endava/internship/mocking/service/BasicValidationServiceTest.java]: ok

File: src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java (hunk 6 above)
Comment (icovdii) [reply src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java]: if not possible to make them final, it is ok, but what about private?

File: src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java (hunk 8 above)
Comment (icovdii) [reply src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java:59]: what will happen if createPayment method is invoked with valid user? Can you cover this case with test?

File: src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java (hunk 11 above)
Comment (icovdii) [reply src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java:71]: What about to assert that message was edited correctly? And that it didn't remain the same after editing? What do you think?

Comment (randy525) [reply src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java:71]: discussed

**Review by icovdii — APPROVED** [review]
Looks good to me.

File: src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java (hunk 8 above)
Comment (icovdii) [reply src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java:59]: ah, sorry, test for happy case exists, I was confused by its name: createPaymentInteractionWithCollaboratorsCheck(), may be more appropriate name will be: testCreatePaymentForExistentUser()?

Comment (randy525) [reply src/test/java/com/endava/internship/mocking/service/PaymentServiceTest.java:59]: yes, it will be better name

**Review by icovdii — APPROVED** [review]

**Pull request merged**